File: backend/src/routes/security.routes.ts
import { Router, Request, Response } from 'express';
import axios from 'axios';
import { URL } from 'url';

const router = Router();

interface ScanResult {
    status: "safe" | "blocked";
    score: number;
    issues: string[];
    domain: string;
    fileSize: number;
    riskLevel: "Low" | "Medium" | "High";
}

const MALICIOUS_PATTERNS = [
    /<script/i,
    /alert\(/i,
    /onerror/i,
    /eval\(/i,
    /SELECT.*FROM/i,
    /INSERT.*INTO/i,
    /DELETE.*FROM/i,
    /DROP.*TABLE/i,
    /UNION.*SELECT/i,
    /OR.*1=1/i,
];

const HARMFUL_KEYWORDS = [
    'drugs', 'bomb', 'terror', 'porn', 'violence', 'hate', 'suicide', 'kill', 'attack'
];

const CSV_INJECTION_PREFIXES = ['=', '+', '-', '@'];

const isPrivateIP = (host: string): boolean => {
    const parts = host.split('.');
    if (parts.length === 4) {
        const first = parseInt(parts[0]);
        const second = parseInt(parts[1]);
        if (first === 10) return true;
        if (first === 127) return true;
        if (first === 172 && (second >= 16 && second <= 31)) return true;
        if (first === 192 && second === 168) return true;
    }
    return host === 'localhost' || host === '::1';
};

router.post('/scan-url', async (req: Request, res: Response) => {
    const { url } = req.body;

    if (!url) {
        return res.status(400).json({ error: 'URL is required' });
    }

    try {
        const parsedUrl = new URL(url);

        // 1. HTTPS validation
        if (parsedUrl.protocol !== 'https:') {
            return res.status(200).json({
                status: "blocked",
                score: 0,
                issues: ["URL must use HTTPS for secure data transfer."],
                domain: parsedUrl.hostname,
                fileSize: 0,
                riskLevel: "High"
            });
        }

        // 2. CSV extension check
        if (!parsedUrl.pathname.toLowerCase().endsWith('.csv')) {
            return res.status(200).json({
                status: "blocked",
                score: 10,
                issues: ["URL must point to a .csv file."],
                domain: parsedUrl.hostname,
                fileSize: 0,
                riskLevel: "High"
            });
        }

        // 3. Private IP check
        if (isPrivateIP(parsedUrl.hostname)) {
            return res.status(200).json({
                status: "blocked",
                score: 0,
                issues: ["Access to private network or localhost is restricted."],
                domain: parsedUrl.hostname,
                fileSize: 0,
                riskLevel: "High"
            });
        }

        // 4. Fetch metadata (File size check)
        let headResponse;
        try {
            headResponse = await axios.head(url, { timeout: 5000 });
        } catch (e) {
            // If HEAD fails, we'll try GET but limit the range or just proceed to GET
        }

        const contentLength = headResponse?.headers['content-length'];
        const fileSize = contentLength ? parseInt(contentLength) : 0;
        const MAX_SIZE = 5 * 1024 * 1024; // 5MB

        if (fileSize > MAX_SIZE) {
            return res.status(200).json({
                status: "blocked",
                score: 20,
                issues: ["File size exceeds 5MB limit."],
                domain: parsedUrl.hostname,
                fileSize: Math.round(fileSize / 1024),
                riskLevel: "High"
            });
        }

        // 5. Content Scanning (First 100KB for safety checks)
        const issues: string[] = [];
        let score = 100;

        try {
            const response = await axios.get(url, {
                timeout: 10000,
                responseType: 'text',
                headers: { 'Range': 'bytes=0-102400' } // Get first 100KB
            });

            const content = response.data;

            // Script/SQL Injection
            for (const pattern of MALICIOUS_PATTERNS) {
                if (pattern.test(content)) {
                    issues.push(`Detected malicious pattern: ${pattern.source}`);
                    score -= 30;
                }
            }

            // Harmful Keywords
            for (const keyword of HARMFUL_KEYWORDS) {
                if (content.toLowerCase().includes(keyword.toLowerCase())) {
                    issues.push(`Detected prohibited keyword: ${keyword}`);
                    score -= 20;
                }
            }

            // CSV Formula Injection
            const lines = content.split('\n');
            let formulaInjections = 0;
            for (const line of lines.slice(0, 100)) { // Check first 100 lines
                const cells = line.split(',');
                for (const cell of cells) {
                    const trimmed = cell.trim();
                    if (CSV_INJECTION_PREFIXES.some(p => trimmed.startsWith(p)) && trimmed.length > 1) {
                        formulaInjections++;
                    }
                }
            }
            if (formulaInjections > 0) {
                issues.push(`Detected ${formulaInjections} potential CSV formula injection attempts.`);
                score -= 25;
            }

        } catch (error) {
            issues.push("Could not verify file content security.");
            score -= 50;
        }

        // Final assessment
        score = Math.max(0, score);
        const riskLevel = score > 80 ? "Low" : score > 40 ? "Medium" : "High";
        const status = score > 60 ? "safe" : "blocked";

        return res.status(200).json({
            status,
            score,
            issues,
            domain: parsedUrl.hostname,
            fileSize: Math.round(fileSize / 1024),
            riskLevel
        });

    } catch (error) {
        return res.status(400).json({ error: 'Invalid URL provided' });
    }
});

export default router;

File: backend_python/app/routes/security_routes.py
from fastapi import APIRouter, HTTPException, Body
import httpx
from urllib.parse import urlparse
import re

router = APIRouter()

MALICIOUS_PATTERNS = [
    re.compile(r'<script', re.I),
    re.compile(r'alert\(', re.I),
    re.compile(r'onerror', re.I),
    re.compile(r'eval\(', re.I),
    re.compile(r'SELECT.*FROM', re.I),
    re.compile(r'INSERT.*INTO', re.I),
    re.compile(r'DELETE.*FROM', re.I),
    re.compile(r'DROP.*TABLE', re.I),
    re.compile(r'UNION.*SELECT', re.I),
    re.compile(r'OR.*1=1', re.I),
]

HARMFUL_KEYWORDS = [
    'drugs', 'bomb', 'terror', 'porn', 'violence', 'hate', 'suicide', 'kill', 'attack'
]

CSV_INJECTION_PREFIXES = ['=', '+', '-', '@']

def is_private_ip(host: str) -> bool:
    parts = host.split('.')
    if len(parts) == 4:
        try:
            first = int(parts[0])
            second = int(parts[1])
            if first == 10: return True
            if first == 127: return True
            if first == 172 and (16 <= second <= 31): return True
            if first == 192 and second == 168: return True
        except:
            pass
    return host == 'localhost' or host == '::1'

@router.post("/scan-url")
async def scan_url(url: str = Body(..., embed=True)):
    if not url:
        raise HTTPException(status_code=400, detail='URL is required')

    try:
        parsed_url = urlparse(url)

        # 1. HTTPS validation
        if parsed_url.scheme != 'https':
            return {
                "status": "blocked",
                "score": 0,
                "issues": ["URL must use HTTPS for secure data transfer."],
                "domain": parsed_url.hostname,
                "fileSize": 0,
                "riskLevel": "High"
            }

        # 2. CSV extension check
        if not parsed_url.path.lower().endswith('.csv'):
            return {
                "status": "blocked",
                "score": 10,
                "issues": ["URL must point to a .csv file."],
                "domain": parsed_url.hostname,
                "fileSize": 0,
                "riskLevel": "High"
            }

        # 3. Private IP check
        if is_private_ip(parsed_url.hostname or ""):
            return {
                "status": "blocked",
                "score": 0,
                "issues": ["Access to private network or localhost is restricted."],
                "domain": parsed_url.hostname,
                "fileSize": 0,
                "riskLevel": "High"
            }

        # 4. Fetch metadata (File size check)
        file_size = 0
        async with httpx.AsyncClient() as client:
            try:
                head_response = await client.head(url, timeout=5.0)
                content_length = head_response.headers.get('content-length')
                file_size = int(content_length) if content_length else 0
            except:
                pass

        MAX_SIZE = 5 * 1024 * 1024 # 5MB

        if file_size > MAX_SIZE:
            return {
                "status": "blocked",
                "score": 20,
                "issues": ["File size exceeds 5MB limit."],
                "domain": parsed_url.hostname,
                "fileSize": round(file_size / 1024),
                "riskLevel": "High"
            }

        # 5. Content Scanning
        issues = []
        score = 100

        try:
            async with httpx.AsyncClient() as client:
                response = await client.get(
                    url, 
                    timeout=10.0,
                    headers={'Range': 'bytes=0-102400'} # Get first 100KB
                )
                content = response.text

                # Script/SQL Injection
                for pattern in MALICIOUS_PATTERNS:
                    if pattern.search(content):
                        issues.append(f"Detected malicious pattern: {pattern.pattern}")
                        score -= 30

                # Harmful Keywords
                for keyword in HARMFUL_KEYWORDS:
                    if keyword.lower() in content.lower():
                        issues.append(f"Detected prohibited keyword: {keyword}")
                        score -= 20

                # CSV Formula Injection
                lines = content.split('\n')
                formula_injections = 0
                for line in lines[:100]:
                    cells = line.split(',')
                    for cell in cells:
                        trimmed = cell.strip()
                        if any(trimmed.startswith(p) for p in CSV_INJECTION_PREFIXES) and len(trimmed) > 1:
                            formula_injections += 1
                
                if formula_injections > 0:
                    issues.append(f"Detected {formula_injections} potential CSV formula injection attempts.")
                    score -= 25

        except Exception as e:
            issues.append("Could not verify file content security.")
            score -= 50

        # Final assessment
        score = max(0, score)
        risk_level = "Low" if score > 80 else "Medium" if score > 40 else "High"
        status = "safe" if score > 60 else "blocked"

        return {
            "status": status,
            "score": score,
            "issues": issues,
            "domain": parsed_url.hostname,
            "fileSize": round(file_size / 1024),
            "risk_level": risk_level
        }

    except Exception as e:
        raise HTTPException(status_code=400, detail='Invalid URL provided')
